### ISSUE TITLE
Always same results

### ISSUE THREAD

**Issue by hakanerten09** [open-mmlab/mmsegmentation#318]
Hi. I'm trying U-net and Resnet. All the time i see the same results. i have 5 classes. First class mıou value nan. Fifth one always 0. The others are changing. Always the same results occured. 

**Example:**
classes = ('forest', 'urban', 'agr_areas', 'others', 'water')
palette = [[127, 127, 127], [1, 1, 1], [63, 63, 63],  [191, 191, 191], [254, 254, 254]]

forest  ıou = nan , acc=nan
urban ıou  =67.08 , acc=74.92
agr_areas ıou= 52.93, acc= 68.59 
 others    ıou=66.35, acc=88.5  
water     ıou=0.0 ,  acc=0.0 

classes = ('others', 'urban', 'water', 'forest', 'agr_areas')
palette = [[191, 191, 191], [1, 1, 1], [254, 254, 254], [127, 127, 127],[63, 63, 63]]

others  ıou = nan , acc=nan
urban ıou  =66.09 , acc=71.25
water ıou= 48.74, acc= 72.45 
forest    ıou=68.59, acc=84.3  
agr_areas     ıou=0.0 ,  acc=0.0 

My u-net code:
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""
import mmcv
# import matplotlib.pyplot as plt
import os.path as osp

# import numpy as np
# from PIL import Image
# import matplotlib.patches as mpatches

"Kendi oluşturduğum dataset"

data_root = 'data/my_dataset'
img_dir = 'img_dir'
ann_dir = 'ann_dir'

classes = ('forest', 'urban', 'agr_areas', 'others', 'water')
palette = [[127, 127, 127], [1, 1, 1], [63, 63, 63],  [191, 191, 191], [254, 254, 254]]

#split train/val set randomly
split_dir = 'splits'
mmcv.mkdir_or_exist(osp.join(data_root, split_dir))
filename_list = [osp.splitext(filename)[0] for filename in mmcv.scandir(
    osp.join(data_root, ann_dir), suffix='.tif')]
with open(osp.join(data_root, split_dir, 'train.txt'), 'w') as f:
  # select first 4/5 as train set
  train_length = int(len(filename_list)*4/5)
  f.writelines(line + '\n' for line in filename_list[:train_length])
with open(osp.join(data_root, split_dir, 'val.txt'), 'w') as f:
  # select last 1/5 as train set
  f.writelines(line + '\n' for line in filename_list[train_length:])

from mmseg.datasets.builder import DATASETS
from mmseg.datasets.custom import CustomDataset

@DATASETS.register_module()
class SarImageDataset(CustomDataset):
    CLASSES = classes
    PALETTE = palette

    def __init__(self, split, **kwargs):
        super().__init__(img_suffix='.tif', seg_map_suffix='.tif',
                         split=split, **kwargs)
        assert osp.exists(self.img_dir) and self.split is not None

from mmcv import Config

cfg = Config.fromfile('configs/Unet_conf.py')

from mmseg.apis import set_random_seed

cfg.model = dict(
    type='EncoderDecoder',
    pretrained=None,
    backbone=dict(
        type='UNet',
        in_channels=3,
        base_channels=64,
        num_stages=5,
        strides=(1, 1, 1, 1, 1),
        enc_num_convs=(2, 2, 2, 2, 2),
        dec_num_convs=(2, 2, 2, 2),
        downsamples=(True, True, True, True),
        enc_dilations=(1, 1, 1, 1, 1),
        dec_dilations=(1, 1, 1, 1),
        with_cp=False,
        conv_cfg=None,
        norm_cfg=dict(type='BN', requires_grad=True),
        act_cfg=dict(type='ReLU'),
        upsample_cfg=dict(type='InterpConv'),
        norm_eval=False),
    decode_head=dict(
        type='FCNHead',
        in_channels=64,
        in_index=4,
        channels=64,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=5,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0)),
    auxiliary_head=dict(
        type='FCNHead',
        in_channels=128,
        in_index=3,
        channels=64,
        num_convs=1,
        concat_input=False,
        dropout_ratio=0.1,
        num_classes=5,
        norm_cfg=dict(type='BN', requires_grad=True),
        align_corners=False,
        loss_decode=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4)))

cfg.dataset_type = 'SarImageDataset'
cfg.data_root = data_root

cfg.data.samples_per_gpu = 8
cfg.data.workers_per_gpu = 8

cfg.img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
cfg.crop_size = (256,256)
cfg.train_pipeline = [  # Training pipeline.
    dict(type='LoadImageFromFile'),  # First pipeline to load images from file path.
    dict(type='LoadAnnotations'),  # Second pipeline to load annotations for current image.
    dict(type='Resize',  # Augmentation pipeline that resize the images and their annotations.
       img_scale=(2048, 512),  # The largest scale of image.
       ratio_range=(0.5, 2.0)), # The augmented scale range as ratio.
    dict(type='RandomCrop',  # Augmentation pipeline that randomly crop a patch from current image.
        crop_size=(256,256),  # The crop size of patch.
        cat_max_ratio=0.75),  # The max area ratio that could be occupied by single category.
    dict(
        type='RandomFlip',  # Augmentation pipeline that flip the images and their annotations
        flip_ratio=0.5),  # The ratio or probability to flip
    dict(type='PhotoMetricDistortion'),  # Augmentation pipeline that distort current image with several photo metric methods.
    dict(
        type='Normalize',  # Augmentation pipeline that normalize the input images
        mean=[123.675, 116.28, 103.53],  # These keys are the same of img_norm_cfg since the
        std=[58.395, 57.12, 57.375],  # keys of img_norm_cfg are used here as arguments
        to_rgb=True),
    dict(type='Pad',  # Augmentation pipeline that pad the image to specified size.
        size=(256,256),  # The output size of padding.
        pad_val=0,  # The padding value for image.
        seg_pad_val=255),  # The padding value of 'gt_semantic_seg'.
    dict(type='DefaultFormatBundle'),  # Default format bundle to gather data in the pipeline
    dict(type='Collect',  # Pipeline that decides which keys in the data should be passed to the segmentor
        keys=['img', 'gt_semantic_seg'])
]
cfg.test_pipeline = [
    dict(type='LoadImageFromFile'),  # First pipeline to load images from file path
    dict(
        type='MultiScaleFlipAug',  # An encapsulation that encapsulates the test time augmentations
        img_scale=(2048, 512),  # Decides the largest scale for testing, used for the Resize pipeline
        flip=False,  # Whether to flip images during testing
        transforms=[
            dict(type='Resize',  # Use resize augmentation
                 keep_ratio=True),  # Whether to keep the ratio between height and width, the img_scale set here will be supressed by the img_scale set above.
            dict(type='RandomFlip'),  # Thought RandomFlip is added in pipeline, it is not used when flip=False
            dict(
                type='Normalize',  # Normalization config, the values are from img_norm_cfg
                mean=[123.675, 116.28, 103.53],
                std=[58.395, 57.12, 57.375],
                to_rgb=True),
            dict(type='ImageToTensor', # Convert image to tensor
                keys=['img']),
            dict(type='Collect', # Collect pipeline that collect necessary keys for testing.
                keys=['img'])
        ])
]

cfg.data.train.type = cfg.dataset_type
cfg.data.train.data_root = cfg.data_root
cfg.data.train.img_dir = img_dir
cfg.data.train.ann_dir = ann_dir
cfg.data.train.pipeline = cfg.train_pipeline
cfg.data.train.split = 'splits/train.txt'

cfg.data.val.type = cfg.dataset_type
cfg.data.val.data_root = cfg.data_root
cfg.data.val.img_dir = img_dir
cfg.data.val.ann_dir = ann_dir
cfg.data.val.pipeline = cfg.test_pipeline
cfg.data.val.split = 'splits/val.txt'

cfg.data.test.type = cfg.dataset_type
cfg.data.test.data_root = cfg.data_root
cfg.data.test.img_dir = img_dir
cfg.data.test.ann_dir = ann_dir
cfg.data.test.pipeline = cfg.test_pipeline
cfg.data.test.split = 'splits/val.txt'

# We can still use the pre-trained Mask RCNN model though we do not need to
# use the mask branch
#cfg.load_from = 'checkpoints/pspnet_r50-d8_512x1024_40k_cityscapes_20200605_003338-2966598c.pth'

# Set up working dir to save files and logs.
cfg.work_dir = './work_dirs/tutorial'

# cfg.total_iters = 200
# cfg.log_config.interval = 10
# cfg.evaluation.interval = 50
# cfg.checkpoint_config.interval = 50
cfg.log_config = dict(
    interval=50, hooks=[dict(type='TextLoggerHook', by_epoch=False)])
cfg.dist_params = dict(backend='nccl')
cfg.log_level = 'INFO'
cfg.load_from = None
cfg.resume_from = None
cfg.workflow = [('train', 1)]
cfg.cudnn_benchmark = True
cfg.optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0005)
cfg.optimizer_config = dict()
cfg.lr_config = dict(policy='poly', power=0.9, min_lr=0.0001, by_epoch=False)
cfg.runner = dict(type='IterBasedRunner', max_iters=160000)
cfg.checkpoint_config = dict(by_epoch=False, interval=16000, max_keep_ckpts=3)
cfg.evaluation = dict(interval=16000, metric='mIoU')

# Set seed to facitate reproducing the result
cfg.seed = 0
set_random_seed(0, deterministic=False)
cfg.gpu_ids = range(1)

# Let's have a look at the final config used for training
print(f'Config:\n{cfg.pretty_text}')
from mmseg.datasets import build_dataset
from mmseg.models import build_segmentor
from mmseg.apis import train_segmentor

# Build the dataset
datasets = [build_dataset(cfg.data.train)]

# Build the detector
model = build_segmentor(
    cfg.model, train_cfg=cfg.train_cfg, test_cfg=cfg.test_cfg)
# Add an attribute for visualization convenience
model.CLASSES = datasets[0].CLASSES

# Create work_dir
mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
train_segmentor(model, datasets, cfg, distributed=False, validate=True,
                meta=dict())

**Comment by Junjun2016** [comment]
Hi @hakanerten09 ,
Do you mean that you got the same results with Resnet-based methods and UNet?
If yes, maybe you should check your data especially the target label.

**Comment by hakanerten09** [comment]
> Hi @hakanerten09 ,
> Do you mean that you got the same results with Resnet-based methods and UNet?
> If yes, maybe you should check your data especially the target label.

hi @Junjun2016 ,
Thanks for quick reply. Answer is No. For UNet for example,

 My classes are;
classes = ('forest', 'urban', 'agr_areas', 'others', 'water')
palette = [[127, 127, 127], [1, 1, 1], [63, 63, 63], [191, 191, 191], [254, 254, 254]]


My first class in the class array IoU result always : nan. (Forest IoU:nan)
It doesn't matter which class it is. Forest,water or urban etc.
Also, My last class in the class array IoU result always : 0. (water IoU:0)
It doesn't matter which class it is again.

I changed the order of class array like;

classes = ('others', 'urban', 'water', 'forest', 'agr_areas')
palette = [[191, 191, 191], [1, 1, 1], [254, 254, 254], [127, 127, 127],[63, 63, 63]]

But result always same.  (Others IoU:nan),( agr_area IoU:0)

Thanks a lot.

**Comment by Junjun2016** [comment]
Hi @hakanerten09 
Please try other methods and check your data.
If you only change the order of class name, the target label will not be changed.
